api/enki.py

```diff
@@ -0,0 +1,542 @@
+        self.library_id = collection.external_account_id.encode("utf8")
+        self.base_url = collection.external_integration.url or self.PRODUCTION_BASE_URL
+
+        if not self.library_id or not self.base_url:
+            raise CannotLoadConfiguration(
+                "Enki configuration is incomplete."
+            )
+        self.enki_bibliographic_coverage_provider = (
+            EnkiBibliographicCoverageProvider(
+                _db, collection, api_class=self
+            )
+        )
+        self.token = "mock_token"
+
+    @property
+    def authorization_headers(self):
+        # I don't think Enki needs authorization headers...
+        #authorization = u":".join([self.username, self.password, self.library_id])
+        authorization = u":".self.library_id
+        authorization = authorization.encode("utf_16_le")
+        authorization = base64.standard_b64encode(authorization)
+        return dict(Authorization="Basic " + authorization)
+
+    def refresh_bearer_token(self):
+        url = self.base_url + self.access_token_endpoint
+        headers = self.authorization_headers
+        response = self._make_request(
+            url, 'post', headers, allowed_response_codes=[200]
+        )
+        return self.parse_token(response.content)
+
+    def request(self, url, method='get', extra_headers={}, data=None,
+                params=None, exception_on_401=False):
+        """Make an HTTP request, acquiring/refreshing a bearer token
+        if necessary.
+        """
+        if not self.token:
+            self.token = self.refresh_bearer_token()
+
+        headers = dict(extra_headers)
+        headers['Authorization'] = "Bearer " + self.token
+        headers['Library'] = self.library_id
+        if exception_on_401:
+            disallowed_response_codes = ["401"]
+        else:
+            disallowed_response_codes = None
+            response = self._make_request(
+            url=url, method=method, headers=headers,
+            data=data, params=params,
+            disallowed_response_codes=disallowed_response_codes
+        )
+        if response.status_code == 401:
+            # This must be our first 401, since our second 401 will
+            # make _make_request raise a RemoteIntegrationException.
+            #
+            # The token has expired. Get a new token and try again.
+            self.token = None
+            return self.request(
+                url=url, method=method, extra_headers=extra_headers,
+                data=data, params=params, exception_on_401=True
+            )
+        else:
+            return response
+
+    def availability(self, patron_id=None, since=None, title_ids=[], strt=0, qty=2000):
+        self.log.debug ("requesting : "+ str(qty) + " books starting at econtentRecord" +  str(strt))
+        url = str(self.base_url) + str(self.availability_endpoint)
+        args = dict()
+	args['method'] = "getAllTitles"
+	args['id'] = "secontent"
```

Comment:
There's a mixture of tabs and spaces here, which can cause subtle bugs. We follow PEP 8 which recommends using spaces instead of tabs, and prohibits mixing tabs and spaces. In another language this wouldn't be a big deal but in Python it's very important. You should be able to configure your editor to use spaces instead of tabs.